### PR TITLE
Fix debug reload

### PR DESCRIPTION
**Pull request by HereThereBeDragons**
switching which signal is send first so that legacy versions can cope with it

### CODE REVIEW

File: cvmfs/loader_talk.cc [hunk 1]
```diff
@@ -66,22 +67,38 @@ static void *MainTalk(void *data __attribute__((unused))) {
...
+            }
+          break;
         }
-      } else if ((command != 'R') && (command != 'S')) {  // legacy support
+      }
+      if (unkown_command) {
         SendMsg2Socket(con_fd, "unknown command\n");
         continue;
-      }
+       }
```
Comment (vvolkl) [inline cvmfs/loader_talk.cc]:
```suggestion
      }
```
whitespace

File: cvmfs/loader_talk.cc [hunk 2]
```diff
@@ -66,22 +67,38 @@ static void *MainTalk(void *data __attribute__((unused))) {
...
+                reload_mode = second_cmd == 'd' ? kReloadDebug : kReloadNoDebug;
+                unkown_command = false;
+              }
+            }
+          break;
         }
-      } else if ((command != 'R') && (command != 'S')) {  // legacy support
+      }
+      if (unkown_command) {
         SendMsg2Socket(con_fd, "unknown command\n");
```
Comment (vvolkl) [inline cvmfs/loader_talk.cc]: printing the unknown command char could be helpful

Comment (HereThereBeDragons) [reply cvmfs/loader_talk.cc]: it is printed thats what `sendmsg2socket` does. (you might remember the error message wie had before `<repo>.cern.ch: unknown command` ... REPO CRASHED etc. that came from there

**Review by vvolkl — APPROVED** [review]

**Review by jblomer — COMMENTED** [review]
@vvolkl @HereThereBeDragons  Please don't tag yet, I think this needs another small change. I'll submit a longer comment later tonight. The short version is: the timeout can trigger for various reasons, so this approach is (too) brittle.

**Review by jblomer — COMMENTED** [review]
Apparently with this fix the order of the two commands gets swapped: first 'S/R' then 'd/n'. 

Is there a reason why this is necessary? In the previous version, the advantage was that the first command already indicated if a new cvmfs or an old cvmfs version sends it. So we don't need the timeout. It should also be backwards compatible, since an old loader would ignore the first `d/n` and then pick up the following `S/R`.

**Comment by jblomer** [comment]
A problem with the previous approach is that we have a spurious "unknown command" message displayed. That should only be a presentation error though. I think we can just announce it in the release notes.

**Comment by HereThereBeDragons** [comment]
it is not only a spurious `unkown command` all follow up calls end with `Reload CRASHED!` etc. It does not seem that the reload is successfully executed.

Why is the timeout not ok? it is 1sec timeout. the 2 cmds written to the pipe are immediately done one after each other. also if the 2nd cmd does not reach, it falls back to the legacy mode: just reloading in the current mode the libraries are. (unless it is an error)

**Comment by jblomer** [comment]
Interesting. The problem is that we don't know for sure that the two commands are sent atomically. I think they are not, because it's two different `WritePipe` operations. Even if it was one, I think there is not a guarantee because the socket is stream oriented and not datagram oriented. So, in between the two commands, the sending process can wait an arbitrary amount of time (it could be swapped out, for instance). That is unlikely, but we have seen in the past that the cvmfs deployment scale is big enough to trigger unlikely events.

It's true though that the worst case effect is limited.  Still, we designed the original approach in a backwards compatible way. And apparently it isn't. So it would be good to understand what we got wrong. Perhaps there is another issue hidden there.

**Pull request merged**